proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,341 @@
+	return rules
+}
+
+// BinaryMemcacheParserFactory implements proxylib.ParserFactory
+type BinaryMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection, injectQueue: make([]queuedInject, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	injectQueue  []queuedInject
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, strData []string, offset uint32) (proxylib.OpType, uint32) {
+	log.Infof("OnData with offset %d", offset)
+
+	//TODO: optimise this
+	data := ([]byte(strings.Join(strData, "")))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	headerMissing := headerSize - len(data)
+
+	if headerMissing > 0 {
+		log.Infof("Did not receive needed header data, need %d more bytes", headerMissing)
+		return proxylib.MORE, uint32(headerMissing)
+	}
+
+	bodyLength := binary.BigEndian.Uint32(data[8:12])
+
+	keyLength := uint32(binary.BigEndian.Uint16(data[2:4]))
+	extrasLength := uint32(data[4])
+
+	if keyLength > 0 {
+		keyMissing := int(headerSize+keyLength+extrasLength) - len(data)
+		if keyMissing > 0 {
+			log.Infof("Did not receive enough bytes for key, need %d more bytes", keyMissing)
+			return proxylib.MORE, uint32(keyMissing)
+		}
+	}
+
+	opcode, key, err := p.parse(data, extrasLength, keyLength)
+	if err != 0 {
+		return proxylib.ERROR, uint32(err)
+	}
+
+	logEntry := &cilium.LogEntry_GenericL7{
+		&cilium.L7LogEntry{
+			Proto: "binarymemcached",
+			Fields: map[string]string{
+				"opcode": strconv.Itoa(int(opcode)),
+				"key":    key,
+			},
+		},
+	}
+
+	// we don't filter reply traffic
+	if reply {
+		log.Infof("reply, passing %d bytes", uint32(len(data)))
+		p.connection.Log(cilium.EntryType_Response, logEntry)
+		p.replyCount++
+		defer p.checkInjectQueue()
+		return proxylib.PASS, uint32(bodyLength + headerSize)
+	}
+
+	p.requestCount++
+
+	matches := p.connection.Matches(binaryMemcacheMeta{opcode, key})
+	if matches {
+		p.connection.Log(cilium.EntryType_Request, logEntry)
+		return proxylib.PASS, bodyLength + headerSize
+	}
+
+	magic := 0x81 | data[0]
+
+	if p.requestCount == p.replyCount+1 {
```

Comment:
Maybe disable the quick path for testing purposes and process all replies through the queue for now?